GLTF2 Import Fixes - Skin(s) to Skeleton - Skin Support

marstaik opened the pull request:
Initial Commit doing quite a bit of work:

- Expanding skins to capture non-joints
- Creating fake joints/stealing nodes for non-joints when they lie in between joints
- Merging overlapping Skins into a Skeleton
- Recomputing Skin IBM's to new Skeleton
- Making MeshInstances use new Skins + Skeletons
- ....and more stuff I don't remember

There needs to be quite a bit of code cleanup still, but functionality wise I think its in a good state and would like additional testing.

Tested GLTF2 files:
[godot-test-proj.zip](https://github.com/godotengine/godot/files/3634256/godot-test-proj.zip)

Preview:
https://f.staik.net/f/2bf.mp4

marstaik committed:
GLTF2 Import Fixes - Skin(s) to Skeleton - Skin Support

marstaik commented:
Note: There is no code added yet to determine if skins happen to be the same, and replace them with one Skin. Currently every gltf_skin creates a Skin in godot.

I can add in some code to remove duplicate Skins and do additional work tomorrow.

fire commented:
Can you test https://github.com/godotengine/godot/pull/29273?

I put some fixes in that pull request you may want to add.

* Set depth draw mode for transparent materials.
* Canonicalize names

fire commented:
I cannot build the branch. See the travis errors.

marstaik commented:
I fixed the crash in the latest update, I had broke something on accident while correcting the ci errors.

marstaik committed:
GLTF Importer - Add more Index types and const-ed up the code

marstaik committed:
More GLTF Fixes

1: Depth draw mode set for transparent materials (iFire)
2: Skeletons
  - Bone names now unique and seperate from scene names
  - Due to mixture of fake joints and joints, new bone sanitizing for names added
  - Fixed an issue where some disjoint skins were not being joined due to a logic error
  - Deterministic and Depth-first bone creation order
3: Skins
  - Removed duplicate skins when possible
4: Animations
  - Fixed invalid morph target names

marstaik commented:
More GLTF Fixes:

1. Depth draw mode set for transparent materials (iFire)

2. Skeletons
  - Bone names now unique and separate from scene names
  - Due to mixture of fake joints and joints, new bone sanitizing for names added
  - Fixed an issue where some disjoint skins were not being joined due to a logic error
  - Deterministic and Depth-first bone creation order
(see Alicia to see the depth-first bone order)

3. Skins
  - Removed duplicate skins when possible
(see Alicia & [Skin_posed_animated_duplicate_skins.zip](https://github.com/godotengine/godot/files/3639380/Skin_posed_animated_duplicate_skins.zip)

4. Animations
  - Fixed invalid morph target names (see Sketchbot)

marstaik committed:
GLTF: Fixed some issues with skin groups joining incorrectly and removed unused code

- Skin groups now merge more cleanly together
- Skins whose highest nodes are siblings of another skin now get merged also
- Skin nodes who have children of another skin now also fuse together
- Removed the re-rooting of IBM code, as it is no longer needed with the Skin system

marstaik commented:
- Fixed some issues with nearby skins not being merged together.
- Removed unused IBM translation code with the new Skin system
- Need to give some Skins identity transforms (not all skins have IBM's)

My hope is that as of this last commit, the gltf importer is in a very, very good state.
(not perfect, but much better 😃 )

akien-mga commented:
Thanks!

aaronfranke commented on core/math/disjoint_set.h:
```diff
@@ -0,0 +1,155 @@
+/*************************************************************************/
+/*  disjoint_set.h                                                       */
```
This file is only used in the GLTF module, does it belong in core?

fire replied:
It is a mathematics algorithm, so belongs in math. @marstaik Thoughts?

aaronfranke replied:
Alright, then we should keep it. It was worth asking the question though :)
